src/wg/color/package.d

```diff
@@ -168,9 +168,168 @@ enum Colors
     yellowGreen          = RGB8(154,205,50)   /// <font color=yellowGreen>&#x25FC;</font>
 }
 
+/**
+ * Create a color from a string.
+ * Params: str = A string representation of a _color.$(BR)
+ * May be a hex _color in the standard forms: (#/$)rgb/argb/rrggbb/aarrggbb$(BR)
+ * May also be the name of any _color from the $(D_INLINECODE Colors) enum.
+ * Returns: The _color expressed by the string.
+ * Throws: Throws $(D_INLINECODE std.conv.ConvException) if the string is invalid.
+ */
+Color colorFromString(Color = RGBA8)(scope const(char)[] str) pure @safe
+{
+    import std.conv : ConvException;
+
+    RGBA8 r;
+    string error = colorFromStringImpl(str, r);
+    if (error)
+        throw new ConvException(error);
+    return r;
+//    return cast(Color)r;
+}
+
+///
+unittest
+{
+    // common hex formats supported:
+
+    // 3 digits
+    assert(colorFromString("F80") == RGB8(0xFF, 0x88, 0x00));
+    assert(colorFromString("#F80") == RGB8(0xFF, 0x88, 0x00));
+    assert(colorFromString("$F80") == RGB8(0xFF, 0x88, 0x00));
+
+    // 6 digits
+    assert(colorFromString("FF8000") == RGB8(0xFF, 0x80, 0x00));
+    assert(colorFromString("#FF8000") == RGB8(0xFF, 0x80, 0x00));
+    assert(colorFromString("$FF8000") == RGB8(0xFF, 0x80, 0x00));
+
+    // 4/8 digita (/w alpha)
+    assert(colorFromString!RGBA8("#8C41") == RGBA8(0xCC, 0x44, 0x11, 0x88));
+    assert(colorFromString!RGBA8("#80CC4401") == RGBA8(0xCC, 0x44, 0x01, 0x80));
+
+    // named colors (case-insensitive)
+    assert(colorFromString("red") == RGB8(0xFF, 0x0, 0x0));
+    assert(colorFromString("WHITE") == RGB8(0xFF, 0xFF, 0xFF));
+    assert(colorFromString("LightGoldenrodYellow") == RGB8(250,250,210));
+
+    // parse failure
+    RGB8 c;
+    assert(colorFromString("Ultraviolet", c) == false);
+}
+
+/**
+ * Create a color from a string.
+ * This version of the function is $(D_INLINECODE nothrow), $(D_INLINECODE @nogc).
+ * Params: str = A string representation of a _color.$(BR)
+ * May be a hex _color in the standard forms: (#/$)rgb/argb/rrggbb/aarrggbb$(BR)
+ * May also be the name of any _color from the $(D_INLINECODE Colors) enum.
+ * color = Receives the _color expressed by the string.
+ * Returns: $(D_INLINECODE true) if a _color was successfully parsed from the string, $(D_INLINECODE false) otherwise.
+ */
+bool colorFromString(Color = RGBA8)(scope const(char)[] str, out Color color) pure nothrow @safe @nogc
+{
+    RGBA8 r;
+    if (colorFromStringImpl(str, r) != null)
+        return false;
+    color = r;
+//    color = cast(Color)r;
+    return true;
+}
+
 
 private:
 
+string colorFromStringImpl(scope const(char)[] str, out RGBA8 color) pure nothrow @safe @nogc
```

Comment:
I guess...?
I mean, this is an `Impl` function, not in the public API, so I'm for whatever makes the code neater.
I actually think this arrangement makes the neatest code. Try and refactor that function as you suggest, I think you'll just have to add a bunch of lines around the place to organise the return value.

Reply:
Step 1: make all preparations
```D
struct Result(T)
{
    private uint msglen;
    private union { T v; immutable(char)* e; }

    bool err() const { return msglen != 0; }

    ref inout(T) val() inout { assert(msglen == 0); return v; }
          string msg() const { assert(msglen != 0); return e[0 .. msglen]; }
}

static assert(Result!uint.sizeof == 16);
static assert(Result!ulong.sizeof == 16);

Result!T Ok(T)(auto ref T val)
{
    Result!T r = { 0, v: val };
    return r;
}

Result!T Err(T)(string msg)
{
    Result!T r = { cast(uint)msg.length, e: &msg[0] };
    return r;
}
```

Step 2: who needs exceptions these days?
```D
Color colorFromString(Color = RGBA8)(scope const(char)[] str)
pure @safe {
    import std.conv : ConvException;

    const r = colorFromStringImpl(str);
    if (r.err)
        throw new ConvException(r.msg);
    else
        return r.val;
}

Result!Color colorFromString(Color = RGBA8)(scope const(char)[] str)
pure nothrow @safe @nogc {
    return colorFromStringImpl(str);
}

Result!RGBA8 colorFromStringImpl(scope const(char)[] str)
pure nothrow @safe @nogc {
    //...
    if (hex)
    {
        //...
        if (hex.length == 3)
        {
            ubyte r = val(hex[0]);
            ubyte g = val(hex[1]);
            ubyte b = val(hex[2]);
            return Ok(RGBA8(
                cast(ubyte)(r | (r << 4)),
                cast(ubyte)(g | (g << 4)),
                cast(ubyte)(b | (b << 4)),
                0));
        }
        //...
        else
            return Err!RGBA8("Invalid length for hex color");
    }
    //...
    static foreach (k; __traits(allMembers, Colors))
    {
        if (streqi(str, k))
        {
            mixin("enum Col = Colors." ~ k ~ ";");
            return Ok(RGBA8(Col.r, Col.g, Col.b, 0xFF));
        }
    }

    return Err!RGBA8("String is not a valid color");
}
```

Reply:
Yep, something like that. I did it similarly with png reader, except I didn't use union and I added `alias v this;`

Reply:
@dayllenger That is literally more code and machinery... what makes that approach simpler?

Reply:
The simplicity appears when you want to return several items, some of which "may be" an error. But in such examples, well, it's a matter of taste. I prefer `Result` over exceptions, it's more functional (the error is one of results) and I don't like `try-catch` blocks.
About performance - [it's not quite true](https://forum.dlang.org/post/q92p9r$1t7q$1@digitalmars.com).

Reply:
Oh I'm certainly not suggesting *exceptions* are faster than `Result` (almost certainly not!), I'm comparing `Result` to what I wrote (using an `out` parameter).

Reply:
Would absolutely NEVER use exceptions anywhere near the kind of code where the perf impact of returning a Result struct might be a disadvantage ;)

Reply:
Actually, after taking another look I think you are right @TurkeyMan. This is the simplest way and since this is internal function and not an API function it is OK to use it here. I actually also didn't use the `Result` with internal png reader functions.

Reply:
It this case it doesn't matter if we have only one `out` parameter. I don't know if the `Result` is faster - maybe, probably, because it is only 16 bytes on 64-bit. I'd prefer it over `out` parameters too, because it's more straitforward.

Reply:
I don't think `Result` is likely to be faster here. It can do small-struct optimisation in this case, but even then it will always break return value optimisation, leading to extra copies in the calling scope, which is a pretty big deal.

Saying it's more straight-forward *might* be true if you believe that it is more straight forward than an out parameter... also if you overlook the machinery necessary to support it, and also the potential perf implications.

I never know the right call to make on this matter... it's a tricky set of trade-offs. If you want to set it as standard, then what occurs is that you need to start making decisions when *not* to use `Result` (because it inhibits return value optimisation), and that starts to get a bit strange; it looks upsetting when you break from the standard because the standard is slow in some cases.

I'm not saying `Result` is a bad pattern, just that it does come with cost, and I don't think we've had any discussion on that yet in the WG, so I've been specifically choosing the 'simplest' least-supporting-machinery choices in cases like this.

If we use something like `Result`, then I think we should use it widely, and that means we need to make choices about exceptions, and work out what the machinery is, and also when it's okay to NOT use `Result`... I didn't feel like we're ready for those discussions yet.

Reply:
Fully agreed with what Manu said.

Reply:
I agree with 'simplest', I only want to add that in such trivial things it is very hard to measure what is faster. I tried to make some benchmarks, but their results are not always in favor of `out`, they are completely undefined and depend on how hard do I push the Enter key...

Reply:
I'm sure it's not substantial in any case, but I'm thinking about things like color objects here, which are objects that tend to be handled in loops that execute *millions* of times, and in complex composite code... I'm nervous that if we inhibit RVO, I don't know what the optimizer will/wont be able to do properly. Chances are, everything inlines, and it's moot as you've probably measured... but I just don't know.

Anyway, it seems like it's about time we had a doc on this and start to think about what we wanna do. I'm actually fairly unemotional here, I don't know what's the best solution.